File: BOJ/2133_1.cpp
// 타일 채우기
// https://www.acmicpc.net/problem/2133
#include <iostream>
using namespace std;
const int MAX = 30 + 1;
int dp[MAX];
/*
    참고
    https://mizzo-dev.tistory.com/entry/baekjoon2133
*/
int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int N;
    cin >> N;

    dp[0] = 1, dp[1] = 0, dp[2] = 3;

    if (N % 2 == 0)
    {
        for (int i = 4; i <= N; i += 2)
        {
            dp[i] = dp[i - 2] * 3;

            for (int j = 4; i - j >= 0; j += 2)
            {
                dp[i] += dp[i - j] * 2;
            }
        }
    }

    cout << dp[N] << '\n';
    return 0;
}

File: BOJ/2133_1.py
'''
    타일 채우기
    https://www.acmicpc.net/problem/2133
'''
import sys
sys.setrecursionlimit(1000000)
input = sys.stdin.readline
MAX = 30 + 1
cache = [-1] * MAX


def tiling(width):
    if width % 2 == 1:
        return 0
    elif width == 0:
        return 1

    if cache[width] != -1:
        return cache[width]

    cache[width] = 3 * tiling(width-2)  # tiling(2) * tiling(width-2)

    i = width-4
    while i >= 0:
        cache[width] += 2 * tiling(i)  # 고유경우 (2) * width[width-4]
        i -= 2
    return cache[width]


# 메인
print(tiling(int(input())))
